### DISCUSSION TITLE
Protecting Python code in Pyodide by Utilizing Cython?

### DISCUSSION THREAD

**Discussion by ehfd** [pyodide/pyodide#916] [Ideas]
https://pyoxidizer.readthedocs.io/en/stable/comparisons.html

There has been many attempts to package Python code mainly in the purpose of protecting the codebase. PyInstaller encrypts .py files using AES256 and hides the key in the executable, making reverse engineering non-trivial. Cython and Nuitka literally compiles Python making code retrieval non-trivial as well. PyOxidizer is in the middle, packing the code into a non-trivial executable.

[This Documentation](https://pyodide.readthedocs.io/en/latest/new_packages.html#manual-creation-of-a-pyodide-package-advanced) seems to only zip the code and produces no additional protection.

What would be a good way to offer code protection (as much as compiled languages) into Pyodide packages in order to make code retrieval from the .js and .data files non-trivial (assuming [This](https://medium.com/@xpl/protecting-python-sources-using-cython-dcd940bb188e) is protective enough)?

**Comment by rth** [comment]
I'm not sure Pyodide is very different from Python running in a classical environment in this respect.

The easiest choice is probably to compile sensitive code (e.g. with Cython) which would also work with Pyodode. 

> seems to only zip the code and produces no additional protection.

I'm not sure protecting them more would help much. With the browser debugger it would be possible to access the decrypted information at runtime. 

Maybe there are some smarter things to try (e.g. writing some custom emscripten FS implementation) but if it's just a module or two compiling might be easier.

**Reply by ehfd** [reply]
@rth How should I generate a .data or .js package from an compiled Cython object (the .so or .pyd equivalent to wasm, not the .c or .pyx files which are basically source code), preferably utilizing the buildpkg workflow?
Will using a setup.py Cython configuration possible to build add the objects to .data and leave out the source code?
I could make a documentation for this if there is a clue to start on.
